src/runtime-rs/crates/hypervisor/src/device_manager.rs

```diff
@@ -0,0 +1,369 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::{
+    utils, BlockDevice, Device, DeviceArgument, GenericConfig, GenericDevice, Hypervisor, IoLimits,
+};
+use agent::types::Device as AgentDevice;
+use anyhow::{anyhow, Result};
+use ini::Ini;
+use kata_sys_util::rand;
+use std::{collections::HashMap, str, sync::Arc};
+use tokio::sync::Mutex;
+/// VirtioMmio indicates block driver is virtio-mmio based
+pub const VIRTIO_MMIO: &str = "virtio-mmio";
+pub const VIRTIO_BLOCK: &str = "virtio-blk";
+pub const VFIO: &str = "vfio";
+const SYS_DEV_PREFIX: &str = "/sys/dev";
+pub const KATA_MMIO_BLK_DEV_TYPE: &str = "mmioblk";
+pub const KATA_BLK_DEV_TYPE: &str = "blk";
+type ArcBoxDevice = Arc<Mutex<Box<dyn Device>>>;
+
+pub struct DeviceManager {
+    block_driver: String,
+    devices: HashMap<String, ArcBoxDevice>,
+    block_index: u64,
+    released_index: Vec<u64>,
+}
+
+impl DeviceManager {
+    pub fn new(block_driver: &str) -> Result<Self> {
+        let driver = match block_driver {
+            VIRTIO_MMIO => VIRTIO_MMIO,
+            // other block driver is not avaliable currently,
+            _ => return Err(anyhow!("Unsupported block driver {}", block_driver)),
+        };
+        Ok(Self {
+            block_driver: String::from(driver),
+            devices: HashMap::new(),
+            block_index: 1,
```

Comment:
block index 1 means virt drive name starts from vdb. Is that correct? We are gonna use pmem as vm rootfs driver. So the guest drive's name will starts from vda. This will cause failure of all mount of block device. May it be better that device manager also manages rootfs' device?

Reply:
Right, the `device manager ` should also manage VM rootfs device. I will fix that.

Reply:
thx

Reply:
It seems that dragonball haven't support virtio-pmem yet. But I need this feature for test right now. So I have to make device-manager accommodate pmem anyway. There is two options:
1. device-manager ignores virtio-pmem entirely. The sandbox inserts pmem rootfs through hypervisor interface directly.
2. device-manager manager pmem device's index and virt path accordingly.

I can help with those. What do you think?

Reply:
I prefer the second option. I think maybe you could fork this branch, and add you implementation on that for testing. This is a large PR, I think it will take a long to be merged into the main

Reply:
OK